docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,471 @@
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This package `flax.optim` contains some optimizers, but that list is far from
+exhaustive and ideally we would instead use JAX optimizers from a dedicated PyPi
+package.
+
+DeepMind already has a dedicated library — [Optax] — that implements a wide
+range of interesting optimizers and provides a framework to compose new
+optimizers from reusable gradient transformations.
+
+[Optax]: https://github.com/deepmind/optax
+
+# Using Optax
+[Using Optax]: #using-optax
+
+## Gradient Transformations
+[Gradient Transformations]: #gradient-transformations
+
+While [Optax] does provide predefined optimizers (like `optax.adam`, or
+`optax.sgd` with momentum), it is really a library of *gradient transformations*
+and the idiomatic way of instantiating an optimizer is by providing a
+combination of these gradient transformations. To emulate the momentum
+optimizer from the example when using the [previous API] we would write:
+
+```python
+import optax
+
+tx = optax.chain(
+    optax.trace(decay=0.9, nesterov=False),
+    optax.scale_by_schedule(lambda step: -get_learning_rate(step)),
+)
+```
+
+Remarks:
+
+- Above gradient transformation would be equivalent with the example under
+  [Optimizer and OptimizerDef] where we define a Momentum optimizer without
+  Nesterov momentum (note that the `beta` parameter corresponds to the `decay`
+  parameter of the `optax.trace()` transformation, and the learning rate is
+  applied in a second chained transformation).
+- Note that hyper parameters like `decay` or `nesterov` only exist in the inner
+  scope of the higher order functions returning the `GradientTransformation`.
+  Such a gradient transformation is currently defined as a `NamedTuple` of the
+  `init()` and the `update()` function. In principle this pattern could be
+  extended to also store hyperparameters, maybe a point to discuss on the
+  [Optax] repo.
+- We can use a `get_learning_rate()` that returns the learning rate depending on
+  the step number when defining the Optax gradient update transformation. Above
+  code illustrates how this can be a drop-in replacement for a function we also
+  use in our [previous training step], where this update function already exists
+  (notice how we need to invert the sign because we add the gradient update to
+  the parameters). In addition, you can use
+  [`inject_hyperparams()`](https://github.com/deepmind/optax/pull/48) to
+  schedule arbitrary hyper parameters with Optax.
+
+## Optax Training Step
+[Optax Training Step]: #optax-training-step
+
+```python
+@functools.partial(jax.jit, static_argnums=(4, 5))
+def train_step(opt_state, variables, inputs, labels, apply_fn, tx_update_fn):
```

Comment:
This would remove two lines from `train_step`, but at the cost of having one more input (now 7 args) and one more return value.

Personally, I find it a bit more natural to think of `variables` in general, and then only split it "close to" the optimizer step where we actually need to treat the individual variable collections differently.

Also note that  this is only necessary with models that update the `model_state`; other models would simply return `params` alone. So with your proposal adding a mutable collection would not only change the signature of `module.apply()`, but also the signature of `train_step()`.

(But I don't have a clear preference here, so I'm happy to update as you proposed if you think it's worth it.)

Reply:
I don't have a strong preference, but I _slightly_ prefer @andsteing's suggestion here. The function args/return list is already quite long for `train_step`, and I think it makes more sense for the signature of `train_step` to stay the same regardless of whether the model has state or not.

I agree with @avital that using fewer Flax specific features would make the code more approachable. Perhaps we can use normal python dict merging here: `{**variables, 'params': params}` instead of `variables.copy(dict(params=params))`?

Reply:
Followed n2cholas's suggestion and replaced calls to `FrozenDict.copy()` with a simple `{**frozen_dict, 'name': value}`. Leaving `train_step()` arguments as-is.

Reply:
LGTM!